### ISSUE TITLE
ML-Benchmark prediction submission: RCMGEM

### ISSUE THREAD

**Issue by henryaddison** [WCRP-CORDEX/ml-benchmark#22]
### Institution ID

UoB

### Institution Name

University of Bristol

### Emulator identifier (`emulator_id`)

RCMGEM-mv-orog

### Emulator Description

RCMGEM is a diffusion model based on work by Song et al. [1], including the hyperparameters (including definision of stochastic differential equation (SDE) that defines the noise-adding process), network architecture and training procedure.

* SDE formulation: sub-Variance-Preserving (``sub-VP")
* Neural network: NCSN++, an improved version of their original Noise Conditional Score Network (NCSN) based on the U-Net architecture used by Ho et al. [2].
* Optimizer: Adam optimizer with learning rate 0.0002.

Uses orography as an input as well as the coarse conditioning variables.

Withheld years 1967, 1975, 2087, 2095 from training set to use as validation set.

Number of epochs depends on domain and training experiment. Epoch chosen for inference based on loss and key metrics (seasonal mean bias, power spectral density, correlation between predictions and RCM truth) on validation subset. Chosen epochs were:

* Emulator Historic Future training experiment:
  * ALPS: 540
  * NZ: 260
  * SA: 560
* ESD pseudo reality training experiment:
  * ALPS: 1000
  * NZ: 300
  * SA: 2000

[1] Song, Yang and Sohl-Dickstein, Jascha and Diederik and Kumar, Abhishek and Ermon, Stefano and Poole, Ben. (2021). "Score-Based Generative Modeling through Stochastic Differential Equations". ICLR. https://openreview.net/forum?id=PxTIG12RRHS
[2] Ho, J., Jain, A., & Abbeel, P. (2020). "Denoising diffusion probabilistic models". NeurIPS.

### Hardware and Training Details

Training and some inference done using ORCHID on JASMIN: One node with one NVIDIA A100 GPU ()

* Training time ~6 minutes per epoch for Emulator_historic_future training experiment and ~3 minutes per epoch for ESD_pseudo_reality training experiment.
* Inference time ~40 minutes per year

Some inference done using Isambard-AI: One node with one H100 Tensor Core GPU (96GB). 

* Inference time ~20 minutes per year.

### Stochastic/Probabilistic Output

yes

### Reference URL

https://arxiv.org/abs/2407.14158

### Repository for Reproducibility

https://zenodo.org/records/18481925

### Additional Notes

I settled on version with orography before the announcement to do both or prioritise without. In my own evaluation there was not a large difference in evaluation performance and possibly required more epochs to converge (not fully explored). I lack time to generate predictions for non-orography.

Linked arxiv paper describing model has recently been accepted by JAMES.

Respository for Reproducibility points to a Zenodo record of the state of the branch of code used for this project (https://github.com/henryaddison/mlde/tree/cordex-ml).

**Comment by jgonzalezab** [comment]
Please fill out [this form](https://forms.gle/1goF329LCrWwdJdh9) so we can give you access to our cluster to transfer your submission.

**Comment by jgonzalezab** [comment]
@henryaddison  I was reviewing the files and noticed that the predictions for `SA_domain/Emulator_hist_future` do not appear to have been submitted. Could you please check whether I may have missed something when processing the ZIP file, or if they were not included on your end?

**Comment by henryaddison** [comment]
Thank you for checking @jgonzalezab. I managed to run the command to add NZ a second time for Emulator_hist_future rather than SA 🤦. I will send it over tonight or in the morning. I'm not sure if rsync will understand if I append to a zip file and instead attempt to resend everything so shall I upload just the `SA_domain/Emulator_hist_future`  predictions as a separate file?

**Comment by jgonzalezab** [comment]
Yes, you can simply upload the `SA_domain/Emulator_hist_future` folder. Please make sure that the resulting ZIP file has a different name from the one originally submitted.

**Comment by jgonzalezab** [comment]
@henryaddison I inspected your files after encountering memory issues while computing diagnostics for your model. I noticed that the NetCDF files include two additional variables:

```
    pr               (member, time, y, x) float64 5GB ...
    raw_pred_pr      (member, time, y, x) float32 2GB ...
    tasmax           (member, time, y, x) float64 5GB ...
    raw_pred_tasmax  (member, time, y, x) float32 2GB ...
```
I believe this is likely the reason your ZIP archive is ~500-600 GB in size. For now, I will ignore the `raw_pred_*` variables, assuming that `pr` and `tasmax` correspond to the final  outputs.

**Comment by henryaddison** [comment]
Ah sorry about that. The `raw_pred_*` variables can indeed be safely ignored. They record what comes directly out of the diffusion model before a simple transform is applied. I added it mainly for debugging the original system to see if it producing a lot of impossible values like large negative rainfall and I should have removed these extra variables.

Currently, uploading the `SA_domain/Emulator_hist_future` as `RCMGEM-mv-orog-SA-Emulator.zip`. Based on first 9% file expected to finish in about 3 hours (so ~17:00 CET/16:00 GMT) though I have found this can fluctuate a lot. Alas this still has the `raw_pred_*` variables and I don't have time to remove them.

**Comment by henryaddison** [comment]
@jgonzalezab Just to confirm the missing predictions finished last night (about 1am). Sorry about both mess ups, thank you for your patience (let's hope some of the predictions are good at least!).

**Comment by jgonzalezab** [comment]
No worries at all and thank you. We completely understand that the submission process isn’t always straightforward.
